refbox/gamemanager.py

```diff
@@ -111,3 +124,80 @@ def timeoutStateRef(self):
 
     def setTimeoutStateRef(self):
         self._timeout_state = TimeoutState.ref
+
+    def addPenalty(self, p):
+        self._penalties[p.team()].append(p)
+
+    def delPenalty(self, p):
+        if p in self._penalties[p.team()]:
+            self._penalties[p.team()].remove(p)
+
+    def penalties(self, team_color):
+        return self._penalties[team_color]
+
+    def deleteAllPenalties(self):
+        self._penalties = [[],[]]
+
+    def _start_unstarted_penalties(self, game_clock):
+        for p in self._penalties[TeamColor.white] + self._penalties[TeamColor.black]:
+            if not p.startTime():
+                p.setStartTime(game_clock)
+
+    def _resetOutstandingPenalties(self):
+        """ Reset any penalties that were interrupted by half-time.
+        """
+        for p in self._penalties[TeamColor.white] + self._penalties[TeamColor.black]:
+            if not p.servedCompletely(0, GameState.first_half):
+                p._duration_remaining = p.timeRemaining(0, GameState.first_half)
+                p._start_time = self._duration
+
+
+class Penalty(object):
+
+    def __init__(self, player, team, duration, start_time = None):
+        self._player = player
+        self._team = team
+
+        # Game time when the penalty started
+        self._start_time = start_time
+
+        # Total time of the penalty
+        self._duration = duration
+
+        # Amount left to be served (might be less than duration if partially
+        # served in the first half)
+        self._duration_remaining = duration
+
+    def setStartTime(self, start_time):
+        self._start_time = start_time
+
+    def startTime(self):
+        return self._start_time
+
+    def timeRemaining(self, game_clock, game_state):
+        if not self._start_time or game_state == GameState.half_time:
+            return self._duration_remaining
+        remaining = self._duration_remaining - (self._start_time - game_clock)
+        return max(remaining, 0)
+
+    def servedCompletely(self, game_clock, game_state):
+        return self.timeRemaining(game_clock, game_state) <= 0
+
+    def player(self):
+        return self._player
+
+    def setPlayer(self, player):
+        self._player = player
+
+    def team(self):
+        return self._team
+
+    def duration(self):
+        return self._duration
+
+    def setDuration(self, duration):
+        self._duration = duration
+        self._duration_remaining = duration
+
+    def dismissed(self):
+        return self._duration == -1
```

Comment:
Is this used?

Reply:
Github is hiding the large diff in ui.py where it's used.